### PR TITLE
Show revamp

### PR DESCRIPTION
**Pull request by ekatek**
Complete rewrite of 'meteor show', with tests. See commit messages for more details.

This is part of the work on the Meteor tool to get longform package descriptions from README.md files into 'meteor show'. It is safe-contained enough, that it makes sense to submit it for review.

(I won't merge to devel until all the README work is complete -- note that this is being pulled to a staging branch).

### CODE REVIEW

File: tools/utils.js [hunk 1]
```diff
@@ -650,6 +649,20 @@ exports.mobileServerForRun = function (options) {
...
+// object and returns a long-form human-readable date (ex: December 9th, 2014).
+exports.longformDate = function (date) {
+  var pubDate = moment(date).format('MMMM Do, YYYY');
+  return pubDate;
+};
+
+// Length of the longest possible string that could come out of longformDate
+// (September is the longest month name, so "September 24th, 2014" would be an
+// example).
+exports.maxDateLength = 20;
```
Comment (ekatek) [inline tools/utils.js]: Heh, I figured that this was more efficient. Will change.

File: tools/commands-packages-query.js [hunk 2]
```diff
@@ -0,0 +1,1159 @@
...
+        constraint: dep.constraint,
+        weak: weak
+      };
+  });
+  return dependencies;
+};
+
+// Converts an object to an EJSON string with the right spacing.
+var convertToEJSON = function (data) {
+  var EJSON = isopackets.load('ejson').ejson.EJSON;
```
Comment (ekatek) [inline tools/commands-packages-query.js]: We have been able to have EJSON isopackets since we separated the former-uniload-catalog out of the remote-catalog BITD! (We couldn't do it before, because we needed to read the remote catalog data in order to load any sort of now-isopacket).

File: tools/commands-packages-query.js [hunk 3]
```diff
@@ -0,0 +1,1159 @@
...
+//    we want that package to show up in our results.
+var getTempContext = function (options) {
+  var projectContext;
+  // If we are running in an app, we will use it to create a
+  // (mostly immutable) projectContext.
+  if (options.appDir) {
+    projectContext = new projectContextModule.ProjectContext({
+      projectDir: options.appDir,
+      // Don't change the state of the app.
+      neverWriteProjectConstraintsFile: true
```
Comment (ekatek) [inline tools/commands-packages-query.js]: Isn't it better to be safe than sorry, in this case? I am afraid of (someone) accidentally running a command on the project that might, under some circumstances, try to modify things, so setting all flags that make sense seems like a good idea as a matter of principle. Is that incorrect? (Once again, not trying to argue, happy to remove extra flags, but I want to be smart in the future, too)

File: tools/commands-packages-query.js [hunk 4]
```diff
@@ -0,0 +1,1159 @@
...
+});
+
+main.registerCommand({
+  name: 'show',
+  pretty: true,
+  minArgs: 1,
+  maxArgs: 1,
+  usesPackage: true,
+  options: {
+    "show-all": { type: Boolean, required: false },
```
Comment (ekatek) [inline tools/commands-packages-query.js]: Wouldn't they notice the error when the option doesn't work and spits out the help anyway? We are doing a fairly large refactor, so deprecating flat-out makes sense, I think. The longer we linger with backwards-compatibility support for something like this, the harder it will be to get rid of it (and the more obnoxious the code will look)

File: tools/commands-packages-query.js [hunk 5]
```diff
@@ -0,0 +1,1159 @@
...
+    // version. In this case, we prefer the local package. Explain our choice to
+    // the user.
+    if (data.local &&
+        catalog.official.getVersion(data.name, data.version)) {
+      Console.info();
+      Console.info(
+        "The same version of this package also exists on the package server.",
+        "To view its metadata, run",
+        Console.command("'meteor show " + data.name + "@" + data.version + "'"),
+        "from outside the project.");
```
Comment (ekatek) [inline tools/commands-packages-query.js]: Yeah.  We thought about that when we were discussing the wording, but decided that anyone using those overrides would already know that. Certainly, mentioning PACKAGE_DIRS in this message seems like too much information and it would be nice to give some hint as to what to do next.

File: tools/commands-packages-query.js (hunk 3 above)
Comment (glasser) [inline tools/commands-packages-query.js]: I mean, meteor currently only edits your project constraints file if you actually write code that calls `projectConstraintsFile.addConstraints` or `projectConstraintsFile.removePackages`.  There's no other magic --- you actually have to hit code that tries to modify a thing called `projectConstraintsFile` in order for the project constraints file to be modified.  For the most part this happens either in a place where you really want the file to be edited (the add or remove commands) or on a project context that's always in a temp directory (in which case it doesn't matter much if the file is written, but it's probably helpful for debugging to write it anyway).

This flag, like the vast majority of options to the ProjectContext constructor, is an antidote for one specific place in all of Meteor where something weird and unusual happens --- we call `projectConstraintsFile.addConstraints` on a possibly non-temporary project but we don't want to change the actual file on disk.  It's weird and has a long and ugly name to alert you the fact that something strange is happening.

I don't think it's appropriate to also use this "SOMETHING WEIRD IS HAPPENING THAT OVERRIDES THE DEFAULTS" flag in cases where the defaults work fine.

Now, maybe this flag was not a great idea in the first place --- maybe this should have been an option to `addConstraints` in the one place where it's relevant.  (And, um, apparently there are two places where it was relevant and I missed one, `--get-ready`, fixing now.)  Though that would make the implementation a bit more complex (what happens if you call addConstraints once with the "don't change on disk" option and once without that option?  should it save some changes and not others?  that's what this API would suggest but it would be unused complexity).

Let me know if you have a better idea about how to design these APIs for special cases.  But with the API as designed now, passing unnecessary flags is counterproductive.

File: tools/commands-packages-query.js [hunk 6]
```diff
@@ -0,0 +1,1159 @@
...
+    }
+
+    // If we have not shown all the available versions, let the user know.
+    if (data.totalVersions > totalShownVersions) {
+      var versionsPluralizer =
+            (data.totalVersions > 1) ?
+            "all " + data.totalVersions + " versions" :
+            "the hidden version";
+      Console.info(
+        "Some versions of", self.name, "have been hidden.",
```
Comment (ekatek) [inline tools/commands-packages-query.js]: It could! I am not sure that it is worth it though? I can't really imagine a common use case where the distinction would matter: no matter what we print, you will either run --show-all next (if you didn't find what you were looking for, and think that you should) or you won't (if you just want an overview). 

For example: I can imagine wanting to know if there is a new, hidden, pre-release version. But knowing that we hid pre-releases doesn't help you, since the hidden pre-releases might have been pre-releases for the current version (ex: <current-version>-rc.0). Besides, for most packages, in the long run, this is almost certainly going to mean 'older versions'.

I am not really against changing this -- maybe removing the ambiguity is exactly the sort of smooth Meteor experience that we want to provide, even if it is not actually useful in practice. Just throwing a counterpoint out there.

If we do opt to remove the ambiguity, how should we deal with old, pre-release versions? My inclination would be to say that old > pre-release That is, we should display the "Older versions of N have been hidden"  if, for example, N is at 10.10.10 and there is a super old 0.0.1-rc.0 hanging around somewhere. That makes sense to me (users should not be alerted to a pre-release version of my package if it is super old). Does it make sense to you?

File: tools/commands-packages-query.js (hunk 3 above)
Comment (glasser) [inline tools/commands-packages-query.js]: (I mean, the answer might be actually something more like "The ProjectConstraintsFile abstraction should be split into two abstractions, one around 'constraints set' and one around 'constraints set from a file written to disk', and you should use the file version only when you want to change the file" or something)

Comment (glasser) [inline tools/commands-packages-query.js]: (I mean, the answer might be actually something more like "The ProjectConstraintsFile abstraction should be split into two abstractions, one around 'constraints set' and one around 'constraints set from a file written to disk', and you should use the file version only when you want to change the file" or something)

File: tools/commands-packages-query.js [hunk 7]
```diff
@@ -0,0 +1,1159 @@
...
+      serverVersionRecords =
+        _.filter(serverVersionRecords, function (vr) {
+          return ! vr.unmigrated && vr.version.indexOf("-") === -1;
+      });
+      serverVersionRecords = _.last(serverVersionRecords, MAX_RECENT_VERSIONS);
+    };
+
+    // Process the catalog records into our preferred format, and look up any
+    // other per-version information that we might need.
+    data["versions"] = _.map(serverVersionRecords, function (versionRecord) {
```
Comment (ekatek) [inline tools/commands-packages-query.js]: Hmm. This is something that we talked about when making the design doc, and decided on this ordering as being more intuitive! I am not super-attached to it either way, but I don't think that we should base this on Atmosphere. For example -- Atmosphere is in a browser, where you start at the top of the page. When (many people) run commands in (many common) terminals, the terminal scrolls to the bottom of the output by default. Hopefully the output of 'meteor show x' can fit in the terminal screen without scrolling, but 'meteor show --show-all x' often can't. It seems to make sense to me that, in that case, you should should be able to see the most recent versions first, and scroll up to see the older ones.

File: tools/commands-packages-query.js (hunk 4 above)
Comment (glasser) [inline tools/commands-packages-query.js]: yeah, sure.  unless you think anyone runs these from scripts.

File: tools/commands-packages-query.js (hunk 6 above)
Comment (glasser) [inline tools/commands-packages-query.js]: I think that last suggestion sounds reasonable.  I dunno, my reaction to the current message really is "WHYYYY?????" :)

File: tools/commands-packages-query.js (hunk 7 above)
Comment (glasser) [inline tools/commands-packages-query.js]: Right, I forgot that some people (ok, most people) use broken terminals that aren't actually text editors with infinite scrollback.  Carry on :)

File: tools/commands-packages-query.js (hunk 4 above)
Comment (ekatek) [inline tools/commands-packages-query.js]: Yeah, but we are also changing the overall output format? So, there is a good chance that they should *probably* consider taking another look at their script anyway! This way, they only have to change their script once.

File: tools/commands-packages-query.js [hunk 8]
```diff
@@ -0,0 +1,1159 @@
...
+  //  - publishedOn: date of publication
+  //  - recommended: true if the version is recommended.
+  _displayAllReleaseVersions: function (versions) {
+    var self = this;
+    var columnOpts = { indent: 2, ignoreWidth: true };
+    // If we don't have any versions, then there is nothing to display.
+    if (! versions) { return; }
+
+    // We are going to print versions with order key ('versions'), separately
+    // from versions without an order key ('experimental versions').
```
Comment (ekatek) [inline tools/commands-packages-query.js]: I think, in this instance, it kind of does? We spent a while thinking about this, which is not to say that we are right. But roughly:
-  we want to order versions with the order key by key! That makes sense. That includes pre-releases (1.0-rc.0) and deprecated versions (1.0). 
-  we also want to display versions without an order key, but we can't order them by key, because they don't have one. A conundrum.
Then, we thought about the circumstances when versions don't have order keys, and decided that it only really covers whacky one-of versions. We could call them 'experimental' versions. That doesn't mean that none of the versions under the normal Versions heading are, strictly-speaking, non-experimental, just that these versions are definitely weird.

File: tools/commands-packages-query.js (hunk 6 above)
Comment (ekatek) [inline tools/commands-packages-query.js]: That's a very sad reaction! I guess I will change it. :)

File: tools/commands-packages-query.js [hunk 9]
```diff
@@ -0,0 +1,1159 @@
...
+
+    if (experimentalVersions) {
+      // We can't sort by order key, so sort by order of publication.
+      experimentalVersions = _.sortBy(experimentalVersions, function (v) {
+        return v.publishedOn;
+      });
+      Console.info("Experimental versions:");
+      var rows = [];
+      _.each(experimentalVersions, function (vr) {
+        // Experimental versions cannot be recommended.
```
Comment (ekatek) [inline tools/commands-packages-query.js]: Noted, but will go in a different CL (since it is a server-side check). :)

File: tools/commands-packages-query.js [hunk 10]
```diff
@@ -0,0 +1,1159 @@
...
+            projectContext.localCatalog.getLatestVersion(pack) ||
+            catalog.official.getLatestMainlineVersion(pack);
+        } else {
+          vr = projectContext.layeredCatalog.getLatestVersion(pack);
+        }
+        if (vr) {
+          matchingPackages.push({
+            name: pack,
+            description: vr.description,
+            latestVersion: vr.version,
```
Comment (ekatek) [inline tools/commands-packages-query.js]: Yes -- do you think that they shouldn't be?

(mmm: ~/meteor/meteor: mmm search --ejson iron:router
{                                             
  "packages": [
    {
      "name": "iron:router",
      "description": "Routing specifically designed for Meteor",
      "latestVersion": "1.0.6",
      "lastUpdated": {
        "$date": 1418959417958
      }
    }
  ],
  "releases": []
})

File: tools/commands-packages-query.js (hunk 4 above)
Comment (glasser) [inline tools/commands-packages-query.js]: agreed

File: tools/commands-packages-query.js (hunk 8 above)
Comment (glasser) [inline tools/commands-packages-query.js]: sounds reasonable.

File: tools/commands-packages-query.js (hunk 10 above)
Comment (glasser) [inline tools/commands-packages-query.js]: no, that makes sense, just sort of talking to myself :)

File: tools/commands-packages-query.js [hunk 11]
```diff
@@ -0,0 +1,1159 @@
...
+});
+
+
+///////////////////////////////////////////////////////////////////////////////
+// search
+///////////////////////////////////////////////////////////////////////////////
+
+main.registerCommand({
+  name: 'search',
+  pretty: true,
```
Comment (glasser) [inline tools/commands-packages-query.js]: Probably should use the just-merged-to-devel "pretty can be a function" support to turn off pretty when ejson is set (for both show and search)

File: tools/help.txt [hunk 12]
```diff
@@ -657,43 +657,41 @@ debugging the Meteor packaging tools themselves.
...
-are known to not be compatible with Meteor 0.9.0 and later. If you
-want to see those packages anyway, use the --show-old option.
-
-By default, meteor search will not show packages that have no official versions
-(ie: non-pre-release versions) If you want to see those packages anyway, use
-the --show-rcs option.
+By default, this will not show packages that have no official versions
+(for example, those that have only prereleases). It will also hide packages
+that, due to migration issues, are known to be incompatible with Meteor 0.9.0
+and later. You can use the show-all option to see hidden packages.
```
Comment (ekatek) [inline tools/help.txt]: Done.

File: tools/main.js [hunk 13]
```diff
@@ -1227,20 +1228,30 @@ Fiber(function () {
...
   }
 
-  if (requiresPackage) {
+  // Some commands have different results when run from a package dir, but don't
+  // strictly require it. These commands should use 'usesPackage' instead of
+  // requiresPackage. (We want to avoid searching up the directory tree for
+  // package.js when we don’t have to. Hopefully, a unified control file will
+  // allow us better control in the future).
+  var usesPackage = command.usesPackage;
+  if (typeof usesPackage === "function") {
```
Comment (ekatek) [inline tools/main.js]: Done.

File: tools/utils.js (hunk 1 above)
Comment (ekatek) [inline tools/utils.js]: Done.

File: tools/commands-packages-query.js [hunk 14]
```diff
@@ -0,0 +1,1159 @@
...
+  }
+  // We don't know when the release was published. Luckily, since there is no
+  // way to use the tool outside of a release, and we always change the tool
+  // between releases, it is a good bet that the release was published on the
+  // same day as the tool.
+  var toolPackage = versionRecord.tool.split('@');
+  var toolName = toolPackage[0];
+  var toolVersion = toolPackage[1];
+  var toolRecord = catalog.official.getVersion(toolName, toolVersion);
+  return new Date(toolRecord.published);
```
Comment (ekatek) [inline tools/commands-packages-query.js]: Done.

File: tools/commands-packages-query.js [hunk 15]
```diff
@@ -0,0 +1,1159 @@
...
+};
+
+// Convert package version dependencies into a more helpful format for output in
+// 'meteor show'.
+var formatDependencies = function (versionDependencies) {
```
Comment (ekatek) [inline tools/commands-packages-query.js]: Done.

File: tools/commands-packages-query.js [hunk 16]
```diff
@@ -0,0 +1,1159 @@
...
+        name: depName,
+        constraint: dep.constraint,
+        weak: weak
+      };
+  });
+  return dependencies;
+};
+
+// Converts an object to an EJSON string with the right spacing.
+var convertToEJSON = function (data) {
```
Comment (ekatek) [inline tools/commands-packages-query.js]: Done.

File: tools/commands-packages-query.js (hunk 3 above)
Comment (ekatek) [inline tools/commands-packages-query.js]: Done.

File: tools/commands-packages-query.js [hunk 17]
```diff
@@ -0,0 +1,1159 @@
...
+    });
+  } else {
+    // We're not in an app, so we will create a temporary app and use it to load
+    // the local catalog. If a local packageDir exists, include it manually.
+    var currentPackageDir = options.packageDir ? [options.packageDir] : [];
+    var tempProjectDir = files.mkdtemp('meteor-show');
+    projectContext = new projectContextModule.ProjectContext({
+      projectDir: tempProjectDir,
+      explicitlyAddedLocalPackageDirs: currentPackageDir,
+      neverWriteProjectConstraintsFile: true,
```
Comment (ekatek) [inline tools/commands-packages-query.js]: Done.

File: tools/commands-packages-query.js [hunk 18]
```diff
@@ -0,0 +1,1159 @@
...
+      neverWritePackageMap: true
```
Comment (ekatek) [inline tools/commands-packages-query.js]: Done.

File: tools/commands-packages-query.js [hunk 19]
```diff
@@ -0,0 +1,1159 @@
...
+//   - showHiddenVersions: return information about all the versions of the
+//     package, including pre-releases and un-migrate versions.
+//   - showDependencies: return information information about
+//     versions' dependencies.
+var PackageQuery = function (options) {
+  var self = this;
+
+  // This is the record in the packages collection. It contains things like
+  // maintainers, and the package homepage.
+  self.metaRecord = options.record;
```
Comment (ekatek) [inline tools/commands-packages-query.js]: Done.

File: tools/commands-packages-query.js [hunk 20]
```diff
@@ -0,0 +1,1159 @@
...
+    // If we haven't found a local record, or if the local record that we found
+    // doesn't match the version that we asked for, then we have to go look in
+    // the server catalog.
+    versionRecord = catalog.official.getVersion(self.name, version);
+    return versionRecord;
+  },
+  // Print the query information to screen.
+  //
+  // options:
+  //   - ejson: Don't pretty-print the data. Return a machine-readable ejson
```
Comment (ekatek) [inline tools/commands-packages-query.js]: Done.

File: tools/help.txt [hunk 21]
```diff
@@ -657,43 +657,41 @@ debugging the Meteor packaging tools themselves.
...
-versions. Use <name>@<version> to get more information about a specific
-version of a package or release.
+Show detailed information on a specific package or release, including
+a list of available versions. Use <name>@<version> to get more information
+about a specific version of a package or release.
 
-By default, meteor show will not show versions that, due to migration issues,
-are known to not be compatible with Meteor 0.9.0 and later. If you
-want to see those versions anyway, use the --show-old option.
+By default, Meteor will not show more than five versions, and will not
```
Comment (ekatek) [inline tools/help.txt]: Done.

File: tools/commands-packages-query.js [hunk 22]
```diff
@@ -0,0 +1,1159 @@
...
+
+    // Processing and formatting architectures takes time, so we don't want to
+    // do this if we don't have to.
+    if (self.showOSArchitectures) {
+      var allBuilds = catalog.official.getAllBuilds(self.name, version);
+      var architectures = _.map(allBuilds, function (build) {
+        return build['buildArchitectures'] &&
+          build.buildArchitectures.split('+')[0];
+      });
+      data["OSarchitectures"] = architectures;
```
Comment (ekatek) [inline tools/commands-packages-query.js]: Done, I think.

File: tools/commands-packages-query.js [hunk 23]
```diff
@@ -0,0 +1,1159 @@
...
+      git: versionRecord.git
+    };
+
+    // Processing and formatting architectures takes time, so we don't want to
+    // do this if we don't have to.
+    if (self.showOSArchitectures) {
+      var allBuilds = catalog.official.getAllBuilds(self.name, version);
+      var architectures = _.map(allBuilds, function (build) {
+        return build['buildArchitectures'] &&
+          build.buildArchitectures.split('+')[0];
```
Comment (ekatek) [inline tools/commands-packages-query.js]: Done.

File: tools/commands-packages-query.js [hunk 24]
```diff
@@ -0,0 +1,1159 @@
...
+        version: version
+      });
+    } catch (e) {
+      // Sometimes, we might be unable to determine if the package is installed
+      // -- maybe we don't have access to the directory, or there is some sort
+      // of disk corruption. This might only extend to one version, so it would
+      // be awkward to fail 'meteor show' altogether. Print an error message (if
+      // it is a permissions error, for example, that's something the user might
+      // want to know), but don't throw.
+      Console.error(e);
```
Comment (ekatek) [inline tools/commands-packages-query.js]: Done.

File: tools/commands-packages-query.js [hunk 25]
```diff
@@ -0,0 +1,1159 @@
...
+    // If we don't have a long-form description, print the summary. (If we don't
+    // have a summary, print nothing).
+    if (data.description || data.summary) {
+      Console.info(data.description || data.summary);
+      Console.info();
+    }
+
+    // Print dependency information, if the package has any dependencies.
+    if (_.has(data, "dependencies") &&
+        ! _.isEmpty(data.dependencies)) {
```
Comment (ekatek) [inline tools/commands-packages-query.js]: Done.

File: tools/commands-packages-query.js [hunk 26]
```diff
@@ -0,0 +1,1159 @@
...
+
+    var totalShownVersions = 0;
+    // If we have any versions to show, print them out now.
+    if (data.versions && ! _.isEmpty(data.versions)) {
+      var versionsHeader =
+            self.showHiddenVersions ? "Versions:" : "Recent versions:";
+      Console.info(versionsHeader);
+      var rows = [];
+      _.each(data.versions, function (v) {
+        totalShownVersions++;
```
Comment (ekatek) [inline tools/commands-packages-query.js]: Done.

File: tools/commands-packages-query.js (hunk 6 above)
Comment (ekatek) [inline tools/commands-packages-query.js]: Done!

File: tools/commands-packages-query.js [hunk 27]
```diff
@@ -0,0 +1,1159 @@
...
+    var data = {
+      track: self.metaRecord.name,
+      maintainers: _.pluck(self.metaRecord.maintainers, "username")
+    };
+    data["defaultVersion"] =
+      catalog.official.getDefaultReleaseVersionRecord(self.name);
+
+    // Collect information about versions.
+    var versions = self.showHiddenVersions ?
+      catalog.official.getReleaseVersionRecords(self.name) :
```
Comment (ekatek) [inline tools/commands-packages-query.js]: Done.

File: tools/commands-packages-query.js [hunk 28]
```diff
@@ -0,0 +1,1159 @@
...
+    "ejson": { type: Boolean, required: false }
+  },
+  catalogRefresh:
+    new catalog.Refresh.OnceAtStart(
+        { maxAge: DEFAULT_MAX_AGE_MS, ignoreErrors: true })
+}, function (options) {
+  // The foo@bar API means that we have to do some string parsing to figure out
+  // if we want a particular version.
+  var itemName = options.args[0];
+  var splitArgs = itemName.split('@');
```
Comment (ekatek) [inline tools/commands-packages-query.js]: Done.

File: tools/commands-packages-query.js (hunk 11 above)
Comment (ekatek) [inline tools/commands-packages-query.js]: Not done, but I'll do that.

**Comment by glasser** [comment]
lgtm

**Comment by ekatek** [comment]
Merged. (Have addressed the various product/wording concerns mentioned by Sashko above either here, or in separate commits)
